File: src/main/java/com/falsepattern/zigbrains/zig/pairing/ZigBraceMatcher.java
package com.falsepattern.zigbrains.zig.pairing;

import com.falsepattern.zigbrains.zig.psi.ZigTypes;
import com.falsepattern.zigbrains.zon.psi.ZonTypes;
import com.intellij.lang.BracePair;
import com.intellij.lang.PairedBraceMatcher;
import com.intellij.psi.PsiFile;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class ZigBraceMatcher implements PairedBraceMatcher {
    public static final BracePair BRACE_PAIR = new BracePair(ZigTypes.LBRACE, ZigTypes.RBRACE, true);
    public static final BracePair PAREN_PAIR = new BracePair(ZigTypes.LPAREN, ZigTypes.RPAREN, false);
    public static final BracePair BRACKET_PAIR = new BracePair(ZigTypes.LBRACKET, ZigTypes.RBRACKET, false);
    private static final BracePair[] PAIRS = new BracePair[]{BRACE_PAIR, PAREN_PAIR, BRACKET_PAIR};
    @Override
    public BracePair @NotNull [] getPairs() {
        return PAIRS;
    }

    @Override
    public boolean isPairedBracesAllowedBeforeType(@NotNull IElementType lbraceType, @Nullable IElementType contextType) {
        return true;
    }

    @Override
    public int getCodeConstructStart(PsiFile file, int openingBraceOffset) {
        var theBrace = file.findElementAt(openingBraceOffset);
        if (theBrace == null) {
            return openingBraceOffset;
        }
        var parent = theBrace.getParent();
        if (parent == null) {
            return openingBraceOffset;
        }
        return parent.getTextOffset();
    }
}

File: modules/core/src/main/kotlin/com/falsepattern/zigbrains/zig/pairing/ZigBraceMatcher.kt
package com.falsepattern.zigbrains.zig.pairing

import com.falsepattern.zigbrains.zig.psi.ZigTypes
import com.intellij.lang.BracePair
import com.intellij.lang.PairedBraceMatcher
import com.intellij.psi.PsiFile
import com.intellij.psi.tree.IElementType

class ZigBraceMatcher: PairedBraceMatcher {
    override fun getPairs() =
        PAIRS

    override fun isPairedBracesAllowedBeforeType(lbraceType: IElementType, contextType: IElementType?) =
        true

    override fun getCodeConstructStart(file: PsiFile?, openingBraceOffset: Int) =
        file?.findElementAt(openingBraceOffset)?.parent?.textOffset ?: openingBraceOffset
}

private val BRACE_PAIR = BracePair(ZigTypes.LBRACE, ZigTypes.RBRACE, true)
private val PAREN_PAIR = BracePair(ZigTypes.LPAREN, ZigTypes.RPAREN, false)
private val BRACKET_PAIR = BracePair(ZigTypes.LBRACKET, ZigTypes.RBRACKET, false)
private val PAIRS = arrayOf(BRACE_PAIR, PAREN_PAIR, BRACKET_PAIR)